forage-android/src/main/java/com/joinforage/forage/android/network/PaymentMethodService.kt

```diff
@@ -43,13 +44,11 @@ internal class PaymentMethodService(
         return convertCallbackToCoroutine(request)
     }
 
-    private fun getPaymentMethodUrl(paymentMethodRef: String): HttpUrl {
-        return httpUrl
-            .newBuilder()
-            .addPathSegment(ForageConstants.PathSegment.API)
-            .addPathSegment(ForageConstants.PathSegment.PAYMENT_METHODS)
-            .addPathSegment(paymentMethodRef)
-            .addTrailingSlash()
-            .build()
-    }
+    private fun getPaymentMethodUrl(paymentMethodRef: String): HttpUrl = httpUrl.toHttpUrlOrNull()!!
```

Comment:
can we refactor this to avoid force unwrapping here?

Reply:
Hey Danilo! I agree with you that we want to squash the risk of a non-null assertion like leading to an accidental runtime error. Here's the situation from what I can see:

The value that get's passed for `httpUrl` is the [`baseUrl` as noted here](https://github.com/teamforage/forage-android-sdk/blob/1048904bea39af315b1e2c566846f5e8c5304be2/sample-app/src/main/java/com/joinforage/android/example/network/model/EnvConfig.kt#L16-L39). To your point, the CI-tests will only surface that if the `staging` URL is broken. Further, local development would likely only surface if `sandbox` or `dev`'s URLs are broken. That means, technically, we don't currently have an automated way of ensuring that the `prod` URL is correct / well-formed.

I'd be happy to write a test [against `EnvConfig`](https://github.com/teamforage/forage-android-sdk/blob/1048904bea39af315b1e2c566846f5e8c5304be2/sample-app/src/main/java/com/joinforage/android/example/network/model/EnvConfig.kt#L11) that ensures that the values of `baseUrl` are correct for each environment. This would give us an automated way of knowing if somebody accidentally broke a URL. However, I don't think outside of an automated test we'd be able to get the compiler's guarantee that a URL will never be null. Allow me to explain:

Here is, for example, an alternative way these get-URLs could be written:
```kotlin
private fun getTokenizeCardUrl(): HttpUrl = Request.Builder()
        .url(httpUrl)
        .build()
        .url
        .newBuilder()
        .addPathSegment(ForageConstants.PathSegment.API)
        .addPathSegment(ForageConstants.PathSegment.PAYMENT_METHODS)
        .addTrailingSlash()
        .build()
```
While the above technically makes the compiler happy that we're working with a non-null type, the [function specs](https://github.com/square/okhttp/blob/c73f9842ad72e4169f2c2f149b5bdee83fb73a76/okhttp/src/jvmMain/kotlin/okhttp3/Request.kt#L191-L196) still say that it will throw an `IllegalArgumentException` if the passed URL is not valid.

Basically the situation seems be that if the URL ever breaks, we'll get a possibly-null variable (with the current implementation) or we can use a function that says it will throw at run time.

My take on take on the best path forward is to add automated tests that ensure we don't break any URLs but let me know what you think!

Reply:
Thank you for the thoughtful insights!

> My take on take on the best path forward is to add automated tests that ensure we don't break any URLs but let me know what you think!

I'm on board with this!

Reply:
Added automated tests in [d505d60](https://github.com/teamforage/forage-android-sdk/pull/126/commits/d505d6079c057423067015ab3e1d2bc18be453df)!

Reply:
Thank you!